[FIX] some entity ajustement

RainproDash opened the pull request:
Update some the entity inventory and project to fit the usage of these. 
Also update the swagger after those modification.

RainproDash committed:
[FIX] some entity ajustement

RainproDash committed:
[FIX] fixing the test and factory also

thmarinho reviewed (changes requested):
On project created, you should items them from 'available', add them back on project completion
Add this to test cases

RainproDash committed:
[ADD] take of the elements out of the inventory when creating a project

RainproDash committed:
[fix] fix the tests, also complete a test to see if the elements are taken out of the inventory after the creation of a project

Tristancalard commented on backend/src/test/projects.test.ts:
```diff
@@ -18,6 +20,7 @@ describe('Project', () => {
                 console.log('Error connecting to database', err);
             });
         repository = dataBase.getRepository(Project);
+        repository2 = dataBase.getRepository(Inventory);
```
same

Tristancalard replied:
same

Tristancalard commented on backend/src/test/projects.test.ts:
```diff
@@ -147,6 +178,14 @@ describe('Project', () => {
+
+        expect(resInv1.body.quantity).to.equal(5);
+        expect(resInv2.body.quantity).to.equal(1);
+        expect(resInv3.body.quantity).to.equal(17);
```
Hard value

Tristancalard replied:
Hard value

Tristancalard commented on backend/src/test/projects.test.ts:
```diff
@@ -3,13 +3,15 @@ import chaiHttp from 'chai-http';
 describe('Project', () => {
 
     let repository: Repository<Project>;
+    let repository2: Repository<Inventory>;
```
Better variable name

RainproDash committed:
[fix] fix the names of variables in tests and hards values

thmarinho commented on backend/src/routes/projects/projects.ts:
```diff
@@ -174,6 +179,11 @@ projectRouter.get('/status/:id', async (req: Request, res: Response) => {
 
 projectRouter.post('/', async (req: Request, res: Response) => {
     try {
+        for (let i: number = 0; i < req.body.elementsIds.length; i++) {
+            const result: any = await dataBase.getRepository(Inventory).createQueryBuilder('inventory').where('inventory.id = :id', { id: req.body.elementsIds[i] }).setParameter('id', String(req.body.elementsIds[i])).getOne();
+            result.quantity = result.quantity - 1;
+            await dataBase.getRepository(Inventory).createQueryBuilder('inventory').update().set({ quantity: result.quantity }).where('inventory.id = :id', { id: req.body.elementsIds[i] }).execute();
+        }
```
make it using .map()

thmarinho commented on backend/src/test/projects.test.ts:
```diff
@@ -3,25 +3,28 @@ import chaiHttp from 'chai-http';
                 console.log('Error connecting to database', err);
             });
-        repository = dataBase.getRepository(Project);
+        repositoryProject = dataBase.getRepository(Project);
```
projectRepository

thmarinho commented on backend/src/routes/projects/projects.ts:
```diff
@@ -174,6 +179,11 @@ projectRouter.get('/status/:id', async (req: Request, res: Response) => {
     try {
+        for (let i: number = 0; i < req.body.elementsIds.length; i++) {
+            const result: any = await dataBase.getRepository(Inventory).createQueryBuilder('inventory').where('inventory.id = :id', { id: req.body.elementsIds[i] }).setParameter('id', String(req.body.elementsIds[i])).getOne();
+            result.quantity = result.quantity - 1;
```
remove quantity received from front not 1

RainproDash replied:
How the body would look like then? Because elementIds is an array of ids
For me if an element was used more than one it would appear several times in the array

RainproDash committed:
[fix]
fix the reduction of elements in the inventory during the creation of a project by using map() insted of a for
fix the format of the body to have the element and his quantity used in a project
fix the swagger to fit those changement
